packages/apollo-federation/src/composition/validate/sdl/uniqueFieldDefinitionNames.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  ASTVisitor,
+  NameNode,
+  GraphQLError,
+  InputObjectTypeDefinitionNode,
+  InterfaceTypeDefinitionNode,
+  ObjectTypeDefinitionNode,
+  InputObjectTypeExtensionNode,
+  InterfaceTypeExtensionNode,
+  ObjectTypeExtensionNode,
+  GraphQLNamedType,
+  isObjectType,
+  isInterfaceType,
+  isInputObjectType,
+} from 'graphql';
+import { SDLValidationContext } from 'graphql/validation/ValidationContext';
+import { TypeMap } from 'graphql/type/schema';
+import Maybe from 'graphql/tsutils/Maybe';
+import { diffTypeNodes } from '../../utils';
+
+type NodeTypesRequiringUniqueFields =
+  | TypeDefinitionsRequiringUniqueFields
+  | TypeExtensionsRequiringUniqueFields;
+
+type TypeDefinitionsRequiringUniqueFields =
+  | InputObjectTypeDefinitionNode
+  | InterfaceTypeDefinitionNode
+  | ObjectTypeDefinitionNode;
+
+type TypeExtensionsRequiringUniqueFields =
+  | InputObjectTypeExtensionNode
+  | InterfaceTypeExtensionNode
+  | ObjectTypeExtensionNode;
+
+export function duplicateFieldDefinitionNameMessage(
+  typeName: string,
+  fieldName: string,
+): string {
+  return `Field "${typeName}.${fieldName}" can only be defined once.`;
+}
+
+export function existedFieldDefinitionNameMessage(
+  typeName: string,
+  fieldName: string,
+): string {
+  return `Field "${typeName}.${fieldName}" already exists in the schema. It cannot also be defined in this type extension.`;
```

Comment:
Is my recommendation accurate?

```suggestion
  return `Field "${typeName}.${fieldName}" already exists in the schema. It cannot also be defined in this type extension. If this is meant to be an external field, add the \`@external\` directive`;
```

Reply:
Also, is serviceName possible here?

Reply:
Yes and yes!
Resolved in 41f04d11